# multiomics
## License
<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.  
## Usage docs  
### Introduction  
The software can be used not only for the analysis of cancer data, but also in any other study where it is necessary to determine the relationship between gene expression and its regulatory factors (in particular, methylation of promoter regions, copy number variations and single nucleotide variations), as well as to determine functional relationships between genes.
The repository contains two files: a Jupyter notebook that runs directly, and an R script that runs inside the notebook.
### Input
As an input for the program, the user must specify the name of the folder containing 4 csv files: gene expression "expr.csv", methylation on gene promoters "prom.csv", CNV "cnv.csv" and SNV "snv.csv". Rows in csv files should be gene names and columns should be sample names.
### Gene clustering
The program works according to the following algorithm. First, those genes whose expression is zero in all samples are removed from the dataset. Then, the genes are grouped sequentially by the following methods:
1. Linear regression. The next method receives only those genes for which it was not possible to identify a significant relationship between the dependent and independent variables (expression and regulating factors). The json file "gene_groups_lm.txt" inside the folder specified by the user contains a list of genes for each regulation group.
2. Generalized linear model with log link function. Gene lists are stored in the json file "gene_groups_glm_log.txt".
3. Generalized linear model with an inverse exponential link function. The lists of genes are stored in the json file "gene_groups_glm_inv_pwr.txt". The list of ungrouped genes is written to the "no_fit_arules.csv" file for later reading inside the R script.
4. Search for association rules. At this stage, the R-script is launched, the input to which is the name of the data folder. To automatically run an R script inside a notebook, it is needed to install the rpy2 library, which allows to run R code directly from cells in a Jupyter notebook. The lists of genes are stored in the json file "gene_groups_arules.txt".
5. Pearson correlation. The lists of genes are stored in the json file "gene_groups_pearson_cor.txt".
As a result of the work of this part of the program, all grouped genes are saved into the json file "gene_groups.txt". Each regulation group is designated by three symbols according to the type of connection of three factors with the level of expression, for example, “–+–”. That is, a negative relationship is denoted as "-", positive - "+". The first character denotes methylation, the second is CNV, and the third is SNV.  
### Functional analysis
In the second part of the program, the functional analysis (gene overrepresentation) is carried out for each group of regulation. Here one can choose from 211 annotated gene set libraries available from Enrichr. By default, 6 libraries are selected: GO_Biological_Process_2021, GO_Cellular_Component_2021, GO_Molecular_Function_2021, Reactome_2022, KEGG_2021_Human, MAGMA_Drugs_and_Diseases. The results of functional analysis are saved as a json file called “enrichr_results.txt”, in which, within each regulation group, for each selected library of annotated gene sets, a list of significantly represented functions and their adjusted p-value will be stored. For clarity, the notebook also displays a comparison of common and unique functions for each regulation group and library. To compare the number of miRNA targets between grouped and ungrouped genes, the program also displays a pie chart and writes the results to the json file “enrichr_results_mir.txt”, which will contain lists of miRNA targets with their corresponding corrected p-values.